feat(sonarr): add indexer methods

Fuochi opened the pull request:
Hi @davidnewhall,
I'm creating this draft to have few feedbacks before continuing. I'm trying to implement the indexer methods for sonarr.
while the /indexer/config endpoint it's pretty straightforward, I'd like to have your opinion on a couple of points for the /indexer endpoint:

- How to address the `Value` field of `Field` struct? The type of this field isn't consistent, should we pass a generic interface? or is it better to have type assertion based on the actual type? I'd prefer to go with the second option. Hence, I tried to implement a draft with the function `correctIndexerFieldsValue` that could be improved if we follow this approach. I even thought of using a generic type assertion switch, but that could lead to wrong assertion for int/float.
- Should we separate input struct and output struct? Like the `Series` struct, the input doesn't need few of the output fields, even if they are natively used by sonarr. While the `Indexer` struct have only 3-4 fields that can be omitted, the main difference will be in the `Field` struct in which the input needs only 2 (`Type` and `Value`) of 11 fields. For now I draft only one struct for both input and output, but I'm not too picky on this one. If you prefer, we can split it up.

What do you think?

p.s. sorry for the lint, I forgot to run it locally. I'll fix it later.

davidnewhall commented:
Is the indexer endpoint documented anywhere?

davidnewhall commented on sonarr/indexer.go:
```diff
@@ -0,0 +1,196 @@
+
+const bpIndexer = APIver + "/indexer"
+
+var (
+	IndexerFieldsString   = []string{"additionalParameters", "apiKey", "apiPath", "baseUrl", "captchaToken", "cookie", "passkey", "username"}
+	IndexerFieldsBool     = []string{"allowZeroSize", "rankedOnly"}
+	IndexerFieldsInt      = []string{"delay", "minimumSeeders", "seedCriteria.seasonPackSeedTime", "seedCriteria.seedTime"}
+	IndexerFieldsFloat    = []string{"seedCriteria.seedRatio"}
+	IndexerFieldsIntSlice = []string{"animeCategories", "categories"}
+)
```
Can't these be struct members?

Fuochi replied:
Don't know. They are the value that `name` parameter should have to map the `value` parameter, I checked them on the source code since I couldn't find docs.

My idea was actually to implement them using a function since we can't have static slices. Something like:
```go
func getIndexerFBool() []string {
	return []string{"allowZeroSize", "rankedOnly"}
}
```

davidnewhall replied:
Thank you for the extra context and json payloads. That helps. 

Questions:
- Is the input data for fields always just name and value? 
- The output data is always _more_ than name and value?

I'm going to think about this for a bit.

_The following is probably not useful yet (I'm still thinking about this, and the below paragraph is just my thoughts that may be accurate)._

What comes to mind first is that we should probably disconnect the input and output struct from the input data. Meaning, we use a different struct to hit sonar than what we expect our library consumers to put into this library. We return a slightly different format for The output. We could even add methods to update these known field types, and perhaps a method that takes in an `interface{}` to catch any we miss or new ones that get added in the future. We could accomplish that with hidden fields that get processed when the payload is passed in, similar to what you're doing now.

Example:
```go
indexer := &Index{Name: "indexer1"}
indexer.SetFieldString("cookie", "string")
indexer.SetFieldBool("allowZeroSize", true)
indexer.SetFieldCategories([]int{})
indexer.SetFieldAnimeCategories([]int{})
```

davidnewhall replied:
I sort of feel like if we know what all these field names (and types) are we should represent them in a `struct`. The struct can be 'processed' into the interface{} data before doing the http request. Reverse that for the output: put the interface{} data back into a struct.

Fuochi commented:
> Is the indexer endpoint documented anywhere?

For sonarr, I haven't found it anywhere. It should be the same as Radarr: https://radarr.video/docs/api/#/Indexer/get_api_v3_indexer__id_
However, they left out the `value` parameter of `fields` which is the main pain point here

Fuochi commented:
input payload (it can work even with less parameters, but it's the way sonarr run the api):
```json
{
    "indexerUrls": [
        "https://thepiratebay.org/",
        "https://piratebayo3klnzokct3wt5yyxb2vpebbuyjl7m623iaxmqhsd52coid.onion.ly/",
        "https://tpb22.ukpass.co/",
        "https://tpb.sadzawka.tk/",
        "https://thepiratebay.host/",
        "https://tpb.cnp.cx/",
        "https://thepiratebay.d4.re/",
        "https://tpb.skynetcloud.site/",
        "https://piratetoday.xyz/",
        "https://piratenow.xyz/",
        "https://pirate-proxy.app/",
        "https://thepb.cyou/",
        "https://mirrorbay.top/",
        "https://5mins.eu/",
        "https://piratebayproxy.buzz/",
        "https://tpb.surf/",
        "https://wearethereal.ninja/",
        "https://tpb.monster/",
        "https://tpb.wtf/",
        "https://bibtor.com/",
        "https://gigatorrent.xyz/",
        "https://thepirateb.xyz/",
        "https://ezrasstuff.com/"
    ],
    "legacyUrls": [
        "https://thepiratebay0.org/",
        "https://thepiratebay10.org/",
        "https://pirateproxy.live/",
        "https://thehiddenbay.com/",
        "https://thepiratebay.zone/",
        "https://tpb.party/",
        "https://piratebayproxy.live/",
        "https://piratebay.live/",
        "https://tpb.biz/",
        "https://pirate.johnedwarddoyle.co.uk/",
        "https://knaben.ru/",
        "https://piratebayztemzmv.onion.pet/",
        "https://piratebayztemzmv.onion.ly/",
        "https://pirateproxy.cloud/",
        "https://tpb18.ukpass.co/",
        "https://pirateproxy.dev/",
        "https://tpb19.ukpass.co/",
        "https://pirateproxy.tube/",
        "https://www.tpbay.win/",
        "https://baypirated.site/",
        "https://piratesbaycc.com/",
        "https://pirateproxy.ltda/"
    ],
    "definitionName": "thepiratebay",
    "description": "The Pirate Bay (TPB) is the galaxy’s most resilient Public BitTorrent site",
    "language": "en-US",
    "enable": true,
    "redirect": false,
    "supportsRss": true,
    "supportsSearch": true,
    "supportsRedirect": false,
    "appProfileId": 1,
    "protocol": "torrent",
    "privacy": "public",
    "capabilities": {
        "limitsMax": 100,
        "limitsDefault": 100,
        "categories": [],
        "supportsRawSearch": false,
        "searchParams": [
            "q"
        ],
        "tvSearchParams": [],
        "movieSearchParams": [],
        "musicSearchParams": [],
        "bookSearchParams": []
    },
    "priority": 25,
    "added": "0001-01-01T00:00:00Z",
    "sortName": "pirate bay",
    "name": "The Pirate Bay",
    "fields": [
        {
            "name": "definitionFile",
            "value": "thepiratebay"
        },
        {
            "name": "baseSettings.queryLimit",
            "value": 100
        },
        {
            "name": "baseSettings.grabLimit",
            "value": 100
        },
        {
            "name": "baseUrl",
            "value": "https://thepiratebay.org/"
        }
    ],
    "implementationName": "Cardigann",
    "implementation": "Cardigann",
    "configContract": "CardigannSettings",
    "infoLink": "https://wiki.servarr.com/prowlarr/supported-indexers#thepiratebay",
    "tags": []
}
```

output payload:
```json
{
    "indexerUrls": [
        "https://thepiratebay.org/",
        "https://piratebayo3klnzokct3wt5yyxb2vpebbuyjl7m623iaxmqhsd52coid.onion.ly/",
        "https://tpb22.ukpass.co/",
        "https://tpb.sadzawka.tk/",
        "https://thepiratebay.host/",
        "https://tpb.cnp.cx/",
        "https://thepiratebay.d4.re/",
        "https://tpb.skynetcloud.site/",
        "https://piratetoday.xyz/",
        "https://piratenow.xyz/",
        "https://pirate-proxy.app/",
        "https://thepb.cyou/",
        "https://mirrorbay.top/",
        "https://5mins.eu/",
        "https://piratebayproxy.buzz/",
        "https://tpb.surf/",
        "https://wearethereal.ninja/",
        "https://tpb.monster/",
        "https://tpb.wtf/",
        "https://bibtor.com/",
        "https://gigatorrent.xyz/",
        "https://thepirateb.xyz/",
        "https://ezrasstuff.com/"
    ],
    "legacyUrls": [
        "https://thepiratebay0.org/",
        "https://thepiratebay10.org/",
        "https://pirateproxy.live/",
        "https://thehiddenbay.com/",
        "https://thepiratebay.zone/",
        "https://tpb.party/",
        "https://piratebayproxy.live/",
        "https://piratebay.live/",
        "https://tpb.biz/",
        "https://pirate.johnedwarddoyle.co.uk/",
        "https://knaben.ru/",
        "https://piratebayztemzmv.onion.pet/",
        "https://piratebayztemzmv.onion.ly/",
        "https://pirateproxy.cloud/",
        "https://tpb18.ukpass.co/",
        "https://pirateproxy.dev/",
        "https://tpb19.ukpass.co/",
        "https://pirateproxy.tube/",
        "https://www.tpbay.win/",
        "https://baypirated.site/",
        "https://piratesbaycc.com/",
        "https://pirateproxy.ltda/"
    ],
    "definitionName": "thepiratebay",
    "description": "The Pirate Bay (TPB) is the galaxy’s most resilient Public BitTorrent site",
    "language": "en-US",
    "encoding": "Unicode (UTF-8)",
    "enable": true,
    "redirect": false,
    "supportsRss": true,
    "supportsSearch": true,
    "supportsRedirect": false,
    "appProfileId": 1,
    "protocol": "torrent",
    "privacy": "public",
    "capabilities": {
        "limitsMax": 100,
        "limitsDefault": 100,
        "categories": [
            {
                "id": 3000,
                "name": "Audio",
                "subCategories": [
                    {
                        "id": 3030,
                        "name": "Audio/Audiobook",
                        "subCategories": []
                    },
                    {
                        "id": 3040,
                        "name": "Audio/Lossless",
                        "subCategories": []
                    },
                    {
                        "id": 3050,
                        "name": "Audio/Other",
                        "subCategories": []
                    },
                    {
                        "id": 3020,
                        "name": "Audio/Video",
                        "subCategories": []
                    }
                ]
            },
            {
                "id": 100100,
                "name": "Audio",
                "subCategories": []
            },
            {
                "id": 100101,
                "name": "Music",
                "subCategories": []
            },
            {
                "id": 100102,
                "name": "Audio Books",
                "subCategories": []
            },
            {
                "id": 100103,
                "name": "Sound Clips",
                "subCategories": []
            },
            {
                "id": 100104,
                "name": "FLAC",
                "subCategories": []
            },
            {
                "id": 100199,
                "name": "Audio Other",
                "subCategories": []
            },
            {
                "id": 2000,
                "name": "Movies",
                "subCategories": [
                    {
                        "id": 2020,
                        "name": "Movies/Other",
                        "subCategories": []
                    },
                    {
                        "id": 2040,
                        "name": "Movies/HD",
                        "subCategories": []
                    },
                    {
                        "id": 2060,
                        "name": "Movies/3D",
                        "subCategories": []
                    }
                ]
            },
            {
                "id": 100200,
                "name": "Video",
                "subCategories": []
            },
            {
                "id": 100201,
                "name": "Movies",
                "subCategories": []
            },
            {
                "id": 100202,
                "name": "Movies",
                "subCategories": []
            },
            {
                "id": 100203,
                "name": "Music Videos",
                "subCategories": []
            },
            {
                "id": 100204,
                "name": "Movie Clips",
                "subCategories": []
            },
            {
                "id": 5000,
                "name": "TV",
                "subCategories": [
                    {
                        "id": 5050,
                        "name": "TV/Other",
                        "subCategories": []
                    },
                    {
                        "id": 5040,
                        "name": "TV/HD",
                        "subCategories": []
                    }
                ]
            },
            {
                "id": 100205,
                "name": "TV",
                "subCategories": []
            },
            {
                "id": 100206,
                "name": "Handheld",
                "subCategories": []
            },
            {
                "id": 100207,
                "name": "HD - Movies",
                "subCategories": []
            },
            {
                "id": 100208,
                "name": "HD - TV shows",
                "subCategories": []
            },
            {
                "id": 100209,
                "name": "3D",
                "subCategories": []
            },
            {
                "id": 100299,
                "name": "Video Other",
                "subCategories": []
            },
            {
                "id": 4000,
                "name": "PC",
                "subCategories": [
                    {
                        "id": 4030,
                        "name": "PC/Mac",
                        "subCategories": []
                    },
                    {
                        "id": 4040,
                        "name": "PC/Mobile-Other",
                        "subCategories": []
                    },
                    {
                        "id": 4060,
                        "name": "PC/Mobile-iOS",
                        "subCategories": []
                    },
                    {
                        "id": 4070,
                        "name": "PC/Mobile-Android",
                        "subCategories": []
                    },
                    {
                        "id": 4050,
                        "name": "PC/Games",
                        "subCategories": []
                    }
                ]
            },
            {
                "id": 100300,
                "name": "Applications",
                "subCategories": []
            },
            {
                "id": 100301,
                "name": "Windows",
                "subCategories": []
            },
            {
                "id": 100302,
                "name": "Mac",
                "subCategories": []
            },
            {
                "id": 100303,
                "name": "UNIX",
                "subCategories": []
            },
            {
                "id": 100304,
                "name": "Handheld",
                "subCategories": []
            },
            {
                "id": 100305,
                "name": "IOS (iPad/iPhone)",
                "subCategories": []
            },
            {
                "id": 100306,
                "name": "Android",
                "subCategories": []
            },
            {
                "id": 100399,
                "name": "Other OS",
                "subCategories": []
            },
            {
                "id": 1000,
                "name": "Console",
                "subCategories": [
                    {
                        "id": 1180,
                        "name": "Console/PS4",
                        "subCategories": []
                    },
                    {
                        "id": 1040,
                        "name": "Console/XBox",
                        "subCategories": []
                    },
                    {
                        "id": 1030,
                        "name": "Console/Wii",
                        "subCategories": []
                    },
                    {
                        "id": 1090,
                        "name": "Console/Other",
                        "subCategories": []
                    }
                ]
            },
            {
                "id": 100400,
                "name": "Games",
                "subCategories": []
            },
            {
                "id": 100401,
                "name": "PC",
                "subCategories": []
            },
            {
                "id": 100402,
                "name": "Mac",
                "subCategories": []
            },
            {
                "id": 100403,
                "name": "PSx",
                "subCategories": []
            },
            {
                "id": 100404,
                "name": "XBOX360",
                "subCategories": []
            },
            {
                "id": 100405,
                "name": "Wii",
                "subCategories": []
            },
            {
                "id": 100406,
                "name": "Handheld",
                "subCategories": []
            },
            {
                "id": 100407,
                "name": "IOS (iPad/iPhone)",
                "subCategories": []
            },
            {
                "id": 100408,
                "name": "Android",
                "subCategories": []
            },
            {
                "id": 100499,
                "name": "Games Other",
                "subCategories": []
            },
            {
                "id": 6000,
                "name": "XXX",
                "subCategories": [
                    {
                        "id": 6010,
                        "name": "XXX/DVD",
                        "subCategories": []
                    },
                    {
                        "id": 6060,
                        "name": "XXX/ImageSet",
                        "subCategories": []
                    },
                    {
                        "id": 6070,
                        "name": "XXX/Other",
                        "subCategories": []
                    }
                ]
            },
            {
                "id": 100500,
                "name": "Porn",
                "subCategories": []
            },
            {
                "id": 100501,
                "name": "Movies",
                "subCategories": []
            },
            {
                "id": 100502,
                "name": "Movies DVDR",
                "subCategories": []
            },
            {
                "id": 100503,
                "name": "Pictures",
                "subCategories": []
            },
            {
                "id": 100504,
                "name": "Games",
                "subCategories": []
            },
            {
                "id": 100505,
                "name": "HD - Movies",
                "subCategories": []
            },
            {
                "id": 100506,
                "name": "Movie Clips",
                "subCategories": []
            },
            {
                "id": 100599,
                "name": "Porn other",
                "subCategories": []
            },
            {
                "id": 0,
                "name": "Other",
                "subCategories": []
            },
            {
                "id": 100600,
                "name": "Other",
                "subCategories": []
            },
            {
                "id": 7000,
                "name": "Books",
                "subCategories": [
                    {
                        "id": 7030,
                        "name": "Books/Comics",
                        "subCategories": []
                    },
                    {
                        "id": 7050,
                        "name": "Books/Other",
                        "subCategories": []
                    }
                ]
            },
            {
                "id": 100601,
                "name": "E-books",
                "subCategories": []
            },
            {
                "id": 100602,
                "name": "Comics",
                "subCategories": []
            },
            {
                "id": 100603,
                "name": "Pictures",
                "subCategories": []
            },
            {
                "id": 100604,
                "name": "Covers",
                "subCategories": []
            },
            {
                "id": 100605,
                "name": "Physibles",
                "subCategories": []
            },
            {
                "id": 100699,
                "name": "Other Other",
                "subCategories": []
            }
        ],
        "supportsRawSearch": false,
        "searchParams": [
            "q",
            "q"
        ],
        "tvSearchParams": [
            "q",
            "season",
            "ep"
        ],
        "movieSearchParams": [
            "q"
        ],
        "musicSearchParams": [
            "q"
        ],
        "bookSearchParams": [
            "q"
        ]
    },
    "priority": 25,
    "added": "2022-06-08T06:24:51Z",
    "sortName": "pirate bay",
    "name": "The Pirate Bay",
    "fields": [
        {
            "order": 0,
            "name": "definitionFile",
            "value": "thepiratebay",
            "type": "textbox",
            "advanced": false,
            "hidden": "hidden"
        },
        {
            "order": 1,
            "name": "baseSettings.queryLimit",
            "label": "Query Limit",
            "helpText": "The number of queries within a rolling 24 hour period Prowlarr will allow to the site",
            "value": 100,
            "type": "number",
            "advanced": true
        },
        {
            "order": 2,
            "name": "baseSettings.grabLimit",
            "label": "Grab Limit",
            "helpText": "The number of grabs within a rolling 24 hour period Prowlarr will allow to the site",
            "value": 100,
            "type": "number",
            "advanced": true
        },
        {
            "order": 3,
            "name": "baseUrl",
            "label": "Base Url",
            "helpText": "Select which baseurl Prowlarr will use for requests to the site",
            "value": "https://thepiratebay.org/",
            "type": "select",
            "advanced": false,
            "selectOptionsProviderAction": "getUrls"
        }
    ],
    "implementationName": "Cardigann",
    "implementation": "Cardigann",
    "configContract": "CardigannSettings",
    "infoLink": "https://wiki.servarr.com/prowlarr/supported-indexers#thepiratebay",
    "tags": [],
    "id": 1
}
```

Fuochi commented:
input payload (it can work even with less parameters but sonarr uses this):
```json
{
    "enableRss": true,
    "enableAutomaticSearch": true,
    "enableInteractiveSearch": true,
    "supportsRss": true,
    "supportsSearch": true,
    "protocol": "torrent",
    "priority": 25,
    "downloadClientId": 0,
    "name": "test",
    "fields": [
        {
            "name": "baseUrl",
            "value": "http://172.17.0.3:9696/1/"
        },
        {
            "name": "apiPath",
            "value": "/api"
        },
        {
            "name": "apiKey",
            "value": "testkey"
        },
        {
            "name": "categories",
            "value": [
                5030,
                5040
            ]
        },
        {
            "name": "animeCategories",
            "value": []
        },
        {
            "name": "animeStandardFormatSearch",
            "value": false
        },
        {
            "name": "additionalParameters"
        },
        {
            "name": "minimumSeeders",
            "value": 1
        },
        {
            "name": "seedCriteria.seedRatio"
        },
        {
            "name": "seedCriteria.seedTime"
        },
        {
            "name": "seedCriteria.seasonPackSeedTime"
        }
    ],
    "implementationName": "Torznab",
    "implementation": "Torznab",
    "configContract": "TorznabSettings",
    "infoLink": "https://wiki.servarr.com/sonarr/supported#torznab",
    "tags": []
}
```
I can't share a valid output right now, but the `fields` parameter element changes to something like:
```json
{
    "pending": false,
    "order": 1,
    "name": "baseSettings.queryLimit",
    "label": "Query Limit",
    "helpText": "The number of queries within a rolling 24 hour period Prowlarr will allow to the site",
    "type": "number",
    "advanced": true,
    "errors": [],
    "warnings": []
}
```

Fuochi commented:
As far as I saw, the input data is always either name and value, or just name if value is empty.
The output data contains all the fields always.

I like your idea of decoupling the user input from the actual api call input, that could work better and also be easier to read.
If we're going to follow this approach, should we have 2 different structs even for user input and output? maybe we can check this during the implementation, when we realize how many fields are actually needed for each.

DownloadClient fields will work in the same way, we should also think if we want the logic to be reusable across resources or if we want to write specific functions for each resource. Maybe we can even implement this logic in starr and manage there all the possible fields for each *arr application, with some default logic for missed ones. Just thought.

btw, I'm not planning to work on this in the next few days. So feel free to add any other comment or suggestion

davidnewhall commented:
> If we're going to follow this approach, should we have 2 different structs even for user input and output? 

For input, yes. One is exported for library consumers, and the other is for calling the actual api in sonar.

For output, I have a couple thoughts.. If we put all the "known fields" into a struct, ie 
```go
type KnownFields struct {
 AdditionalParameters *string
 AllowZeroSize        *bool
 Delay                *int
 AnimeCategories      []int
 ...
 Others               interface{}
}
```
we could probably share this between the input and output. Perhaps methods on this struct make sense too? I'm not sure, but it's an idea. The methods could be like:
```
func (k *KnownFields) Get<FieldName>() (<FieldType>, error) {}
func (k *KnownFields) Set<FieldName>(val <FieldType>) error {}
```
This is just a thought and may not really be useful or needed.

Fuochi commented:
Hi @davidnewhall,
I tried to implement few of your suggestion with this commit (still at very early stage).
the main points are:

- In shared resources I tried to keep the `Field` struct, that could be reused for indexer and download clients, there I added a couple of methods `GetFieldValueString` and `GetFieldValueBool` that could be used during the implementation.
- In indexer, I split the user's input and output, and I decoupled it from the struct used for the api. 
- Also, I created the `IndexerFields` struct to be used by users.
- I'm still not sure on how we could manage the actual fields. I used the two functions `getIndexerFields` and `setIndexerFields` to manage them, but I don't like the idea of using pointers in the `IndexerFields`. Still, I had to use them, at least for booleans, because in `setIndexerFields` the booleans will need to be set for both `true` and `false` values, while skipped if they're not present, hence I can't do something like: 
  ```go
  if fields.AllowZeroSize != false {
  		output = append(output, &starr.Field{
  			Name:  "allowZeroSize",
  			Value: fields.AllowZeroSize,
  		})
  	}
  ```
  because the `false` value is both the zero value for `AllowZeroSize`, and an acceptable value for us. I'm not sure If I managed to explain myself within this point, sorry!

davidnewhall commented on shared.go:
```diff
@@ -118,6 +119,43 @@ type Value struct {
 }
 
+// Field is generic Name/Value struct applied to a few places.
+type Field struct {
```
It might make sense to call this `OutputField` since it will never be an input.

Fuochi replied:
Right now, this is both input and output, I can either split them in `OutputField` and `InputField`, and keep 2 different `IndexerInput` and `IndexerOutput`, or have just one `Indexer` and one `Field`. Surely now it's inconsistent and must be changed.

davidnewhall replied:
I didn't realize it was both. nbd then.

davidnewhall commented on shared.go:
```diff
@@ -118,6 +119,43 @@ type Value struct {
+	if v, a := f.Value.(string); !a {
+		return v, nil
+	}
+	return "", fmt.Errorf("Field %s has invalid value", f.Name)
```
There's a couple problem with this approach. Since this is a dynamic error, the caller can never use `errors.Is` to test it. The caller could just do the type assertion themselves, and handle it however they want anyway, which makes the usefulness of this method questionable. I know I recommended this, but as I see it now it seems less useful.

Fuochi replied:
Yes, I struggled a bit with this one. I think it is needed if we define the fields as user struct output, whereas it's not if we use a generic interface... For sure if we keep the interface, and avoid all this, the SDK implementation will be easier and cleaner. Also, a meaningful error will still be sent back in API response. However, we'll burden the user with the type assertion, and they must know which fields are required and which are optional (which is fine for me). If you prefer to proceed this way, It's fine for me.

davidnewhall replied:
> they must know which fields are required and which are optional 

Indeed, and this is sort of true regardless of how we implement our interface.

Fuochi replied:
Ok, let's keep that easy and clean. I'll change this accordingly

davidnewhall reviewed (commented):
I dunno man. The more I look at this, the more I think `value` should just be an `interface{}`. Can you describe an example of where requiring an input or providing an output with these values as an `interface{}` is a problem? It sort of feels like you're trying to do type checking on types that belong to Radarr/Sonarr; when maybe you can just pass the raw data and let the starr app do the type checking.

Fuochi commented on sonarr/indexer.go:
```diff
@@ -0,0 +1,234 @@
+	Tags                    []*starr.Value `json:"tags"`
+}
+
+type IndexerOutput struct {
```
That was my initial approach, I thought you preferred to keep them separated.

davidnewhall replied:
I thought your initial approach was 1 struct? Maybe I'm misremembering.

Fuochi replied:
yes, my initial approach is one struct only. If it's fine for you, I'll go back to that approach.

davidnewhall replied:
I'm suggesting embedding the input struct into the output struct. It's an idea right now, and not something I think you need to do, but something we should consider.

Like this:
```
type IndexerOutput struct {
 	IndexerInput
 	ImplementationName      string         `json:"implementationName"`
 	InfoLink                string         `json:"infoLink"`
 	SupportsRss             bool           `json:"supportsRss"`
 	SupportsSearch          bool           `json:"supportsSearch"`
}
```

Fuochi replied:
Ok, got it. It really depends on the `Field` struct, if we want to have a `FieldInput` and a `FieldOutput` struct, we'll need to keep `IndexerInput` and `IndexerOutput` separated, otherwise we can do as you suggested.

davidnewhall replied:
I suppose I'm going to recommend splitting it all up then. The inputs and outputs are just different enough that it makes sense to decouple them.

Fuochi commented:
> I dunno man. The more I look at this, the more I think `value` should just be an `interface{}`. Can you describe an example of where requiring an input or providing an output with these values as an `interface{}` is a problem? It sort of feels like you're trying to do type checking on types that belong to Radarr/Sonarr; when maybe you can just pass the raw data and let the starr app do the type checking.

I changed it to have a simple interface. when you find time, have a look. If we agree on the approach, I'll fix the build, write the test, and convert to actual PR

davidnewhall reviewed (commented):
Really like this now. Thank you for going through this exercise with me!

Fuochi commented on shared.go:
```diff
@@ -118,6 +118,35 @@ type Value struct {
 }
 
+// FieldOutput is generic Name/Value struct applied to a few places.
+type FieldOutput struct {
```
They should, when I started working at this I think I checked sonarr, radarr, and prowlarr. But I could doublecheck. Meanwhile I'll fix this PR.

bakerboy448 commented:
looks like you may be missing seed goals?

also something that may help would be to leverage Prowlarr's logic as insporation as it already uses Sonarr's API
https://github.com/Prowlarr/Prowlarr/blob/develop/src/NzbDrone.Core/Applications/Sonarr/SonarrIndexer.cs

davidnewhall commented:
seed goals are probably one of the Fields. Not every indexer has seed goals. Those are dynamic.

Fuochi commented:
Correct, in the first draw of this PR, we were actually trying to map those fields 1-1 (and doing some checks on input as well).
Here the reference:
https://github.com/golift/starr/blob/86e51f711d92d6ee2424e86ea8b2fa7035ce7a08/sonarr/indexer.go#L46-L64
However, we decided to keep Fields generic, and let that part being managed by end user directly.

davidnewhall commented:
I'm going to rewrite some things after you merge, so give me a week before you start your next PR please.
